### PR TITLE
add custom href generator.

### PR DESCRIPTION
**Pull request by vwxyzh**
@chenkennt @fenxu @vicancy @hellosnow @ansyral @qinezh @superyyrrzz

### CODE REVIEW

File: src/Microsoft.DocAsCode.Plugins/IHrefInfo.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+namespace Microsoft.DocAsCode.Plugins
+{
+    public interface IHrefInfo
+    {
+        string FileInSource { get; }
+        string FileInDest { get; }
+        string TargetFileInSource { get; }
+        string TargetFileInDest { get; }
+        string TargetFileFromSourceFile { get; }
+        string DefaultHref { get; }
```
Comment (fenxu) [inline src/Microsoft.DocAsCode.Plugins/IHrefInfo.cs]: OriginalTargetFileLink
DefaultResolvedTargetFileLink

File: src/Microsoft.DocAsCode.Plugins/IHrefInfo.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Plugins
+{
+    public interface IHrefInfo
+    {
+        string FileInSource { get; }
+        string FileInDest { get; }
+        string TargetFileInSource { get; }
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Plugins/IHrefInfo.cs]: >TargetFileInSource [](start = 15, length = 18)

the naming confuses me and I am unable to figure out what they mean from names....

Comment (vicancy) [inline src/Microsoft.DocAsCode.Plugins/IHrefInfo.cs]: What is FileInSouce and TargetFileInSource... why Source contains Target File.... Is any file in source SourceFile?

---
In reply to: [95518402](https://github.com/dotnet/docfx/pull/1146#discussion_r95518402) [](ancestors = 95518402)

File: src/Microsoft.DocAsCode.Plugins/IHrefInfo.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+        string TargetFileInDest { get; }
+        string TargetFileFromSourceFile { get; }
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Plugins/IHrefInfo.cs]: >TargetFileFromSourceFile [](start = 15, length = 24)

FileToTargetFileInSource?

File: src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs [hunk 4]
```diff
@@ -263,7 +263,7 @@ private static void TransformHtml(IDocumentBuildContext context, string html, st
             }
         }
 
-        private static List<CrossReferenceNotResolvedException> TransformHtmlCore(IDocumentBuildContext context, string sourceFilePath, string destFilePath, HtmlAgilityPack.HtmlDocument html)
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs]: >static [](start = 16, length = 6)

why not `static`?

**Comment by superyyrrzz** [comment]
:shipit:

**Comment by qinezh** [comment]
:shipit:

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**